File: planner/src/main/java/com/cozy/planner/config/SecurityConfig.java
package com.cozy.planner.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpStatus;
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.oauth2.core.oidc.user.OidcUser;
import org.springframework.security.web.server.SecurityWebFilterChain;
import org.springframework.security.web.server.authentication.logout.RedirectServerLogoutSuccessHandler;
import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatchers;

import java.net.URI;

@Configuration
@EnableWebFluxSecurity
public class SecurityConfig {

    @Bean
    public SecurityWebFilterChain securityWebFilterChain(ServerHttpSecurity http) {
        var logoutHandler = new RedirectServerLogoutSuccessHandler();
        logoutHandler.setLogoutSuccessUrl(URI.create("/signin"));

        return http
                .authorizeExchange(exchanges -> exchanges
                        .pathMatchers("/css/**", "/js/**", "/signin", "/login**", "/demo-login", "/k6-login", "/oauth2/**", "/webjars/**", "/error**", "/api/v1/events", "/api/v1/ws", "/trainee/**", "/api/v1/trainee/**", "/api/v1/trainee/invite", "/api/v1/me", "/api/v1/trainees/*/availability", "/api/v1/notifications/**", "/api/v1/push/**", "/api/v1/telegram/webhook", "/api/v1/telegram/webhook/**", "/api/v1/telegram/config", "/actuator/**", "/shared/**", "/api/v1/shared/**")
                        .permitAll()
                        .anyExchange().authenticated()
                )
                .oauth2Login(oauth2 -> oauth2
                        .authenticationSuccessHandler((webFilterExchange, authentication) -> {
                            var principal = (OidcUser) authentication.getPrincipal();
                            var session = webFilterExchange.getExchange().getSession();
                            return session.flatMap(s -> {
                                s.getAttributes().put("google_sub", principal.getSubject());
                                s.getAttributes().put("user_email", principal.getEmail());
                                s.getAttributes().put("user_name", principal.getGivenName());
                                var response = webFilterExchange.getExchange().getResponse();
                                response.setStatusCode(org.springframework.http.HttpStatus.FOUND);
                                response.getHeaders().setLocation(URI.create("/setup"));
                                return response.setComplete();
                            });
                        })
                )
                .logout(logout -> logout
                        .logoutSuccessHandler(logoutHandler)
                        .requiresLogout(ServerWebExchangeMatchers.pathMatchers("/logout"))
                )
                .exceptionHandling(handling -> handling
                        .authenticationEntryPoint((exchange, e) -> {
                            var response = exchange.getResponse();
                            response.getHeaders().setLocation(URI.create("/signin"));
                            response.setStatusCode(HttpStatus.FOUND);
                            return response.setComplete();
                        })
                )
                .csrf(ServerHttpSecurity.CsrfSpec::disable)
                .build();
    }
}

File: linkease/server/src/main/kotlin/com/linkease/server/config/SecurityConfig.kt
package com.linkease.server.config

import jakarta.servlet.http.HttpServletResponse
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.oauth2.core.oidc.user.OidcUser
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import org.springframework.security.web.authentication.DelegatingAuthenticationEntryPoint
import org.springframework.security.web.authentication.LoginUrlAuthenticationEntryPoint
import org.springframework.security.web.util.matcher.AntPathRequestMatcher

// Single-tenant login: only app.owner-email may use this app. Mirrors planner's Google
// OAuth2-login shape (same secret env var names so the same secrets file values work),
// but with its own session store/code — planner itself is never touched.
@Configuration
@EnableWebSecurity
class SecurityConfig(private val appProperties: AppProperties) {

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        val apiUnauthorizedEntryPoint = AuthenticationEntryPointStatus(HttpServletResponse.SC_UNAUTHORIZED)
        val entryPointResolver = DelegatingAuthenticationEntryPoint(
            linkedMapOf<org.springframework.security.web.util.matcher.RequestMatcher, org.springframework.security.web.AuthenticationEntryPoint>(
                AntPathRequestMatcher("/api/**") to apiUnauthorizedEntryPoint,
            ),
        ).apply {
            setDefaultEntryPoint(LoginUrlAuthenticationEntryPoint("/oauth2/authorization/google"))
        }

        return http
            .csrf { it.disable() }
            .authorizeHttpRequests { auth ->
                auth
                    .requestMatchers(
                        "/", "/index.html", "/*.js", "/*.wasm", "/*.css", "/*.ico",
                        "/oauth2/**", "/login/**", "/error", "/actuator/health",
                        "/api/me", "/api/telegram/webhook/**",
                    ).permitAll()
                    .anyRequest().authenticated()
            }
            .exceptionHandling { it.authenticationEntryPoint(entryPointResolver) }
            .oauth2Login { oauth2 ->
                oauth2.successHandler(ownerOnlySuccessHandler())
                oauth2.failureHandler { _, response, _ -> response.sendRedirect("/?error=login_failed") }
            }
            .logout { it.logoutSuccessUrl("/") }
            .build()
    }

    private fun ownerOnlySuccessHandler() = AuthenticationSuccessHandler { request, response, authentication ->
        val email = (authentication.principal as OidcUser).email
        if (appProperties.ownerEmail.isBlank() || email != appProperties.ownerEmail) {
            request.session.invalidate()
            response.sendRedirect("/?error=unauthorized")
        } else {
            response.sendRedirect("/")
        }
    }
}

private class AuthenticationEntryPointStatus(
    private val status: Int,
) : org.springframework.security.web.AuthenticationEntryPoint {
    override fun commence(
        request: jakarta.servlet.http.HttpServletRequest,
        response: HttpServletResponse,
        authException: org.springframework.security.core.AuthenticationException,
    ) {
        response.status = status
    }
}
